src/storage/src/source/postgres/snapshot.rs

```diff
@@ -197,23 +197,60 @@ pub(crate) fn render<G: Scope<Timestamp = MzOffset>>(
 
     let is_snapshot_leader = config.responsible_for("snapshot_leader");
 
-    // A filtered table info containing only the tables that this worker should read
-    let mut reader_table_info = table_info.clone();
-    reader_table_info.retain(|oid, _| config.responsible_for(oid));
+    // Determine which exports need to be snapshot, i.e. their offsets are 0.
+    let mut exports_to_snapshot: BTreeSet<_> = export_resume_uppers
+        .into_iter()
+        .filter_map(|(id, upper)| {
+            if upper.into_option() == Some(MzOffset::minimum()) {
+                // export_resume_uppers includes the remap collection, which source exports does
+                // not, which is fine, hence the map.
+                config.source_exports.get(&id).map(|e| e.output_index)
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    // A filtered table info containing only the tables that this worker should snapshot. We filter
+    // this out in the worker itself because we have multiple dependencies on
+    // `exports_to_snapshot`'s state and need to linearize them.
+    let mut reader_snapshot_table_info = table_info.clone();
+
     let (button, errors) = builder.build_fallible(move |caps| {
         Box::pin(async move {
             let id = config.id;
             let worker_id = config.worker_id;
 
             let [data_cap, rewind_cap, snapshot_cap]: &mut [_; 3] = caps.try_into().unwrap();
             let data_cap = data_cap.as_mut().unwrap();
-            trace!(%id, "timely-{worker_id} initializing table reader {}", resume_upper.pretty());
-
-            // We have already snapshotted, so the replication dataflow will take over
-            if resume_upper.into_option() != Some(MzOffset::from(0)) {
+            trace!(
+                %id,
+                "timely-{worker_id} initializing table reader with {} and {} tables to snapshot",
+                config.resume_upper.pretty(),
+                reader_snapshot_table_info.len()
+            );
+
+            // Nothing needs to be snapshot.
+            if exports_to_snapshot.is_empty() {
                 return Ok(());
             }
 
+            // Determine which tables this worker's responsible for that need to be snapshot.
+            reader_snapshot_table_info.retain(|oid, (output_index, _, _)| {
+                mz_ore::soft_assert!(
+                    *output_index != 0,
+                    "primary collection should not be represented in table info"
+                );
+                exports_to_snapshot.remove(output_index) && config.responsible_for(oid)
```

Comment:
Am I reading this right that we unconditionally remove all output_indices regardless of whether or not we're responsible for it?

Reply:
From `exports_to_snapshot` (the global list of everything that needs to be snapshot), yes. We do that so we can understand which table to retain for each worker's snapshot, while also being able to make a simple assertion below.

Reply:
I see. It's a bit weird that this side effect is included in a the predicate of a `retain` that computes a worker-local thing. The invariant that this is asserting has to do with the well-formedness of `table_info` which is constructed on level up. Do you mind moving the assertion in the `PostgresSourceConnection::render` function near where the table info is made?